// mast-api/dist/mast-api.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import axios from "axios";
import { createWriteStream, mkdir, readFile, writeFile } from "fs";
import { RequestFormat, } from "./interfaces.js";
// TODO refactor interfaces and enums into files
export var ObservationCollection;
(function (ObservationCollection) {
    ObservationCollection["HLSP"] = "HLSP";
    ObservationCollection["SPITZER_SHA"] = "SPITZER_SHA";
    ObservationCollection["HST"] = "HST";
    ObservationCollection["TESS"] = "TESS";
    ObservationCollection["JWST"] = "JWST";
    // TODO rest
})(ObservationCollection || (ObservationCollection = {}));
export var Service;
(function (Service) {
    Service["NameLookup"] = "Mast.Name.Lookup";
    Service["ConeSearch"] = "Mast.Caom.Cone";
    Service["Products"] = "Mast.Caom.Products";
    Service["Filtered"] = "Mast.Caom.Filtered";
    Service["FilteredPosition"] = "Mast.Caom.Filtered.Position";
})(Service || (Service = {}));
export class MastApi {
    static makeDir(dir) {
        return __awaiter(this, void 0, void 0, function* () {
            yield new Promise((resolve, reject) => {
                mkdir(dir, (error) => {
                    if (error) {
                        reject(error);
                    }
                    else {
                        resolve();
                    }
                });
            });
        });
    }
    static makeCacheDir() {
        return __awaiter(this, void 0, void 0, function* () {
            yield MastApi.makeDir(MastApi.CACHE_DIR);
        });
    }
    static makeDataDir() {
        return __awaiter(this, void 0, void 0, function* () {
            yield MastApi.makeDir(MastApi.DATA_DIR);
        });
    }
    static writeFile(path, data, isRetry = false) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                yield new Promise((resolve, reject) => {
                    writeFile(path, data, (error) => {
                        if (error) {
                            reject(error);
                        }
                        else {
                            resolve();
                        }
                    });
                });
            }
            catch (error) {
                if (isRetry) {
                    throw error;
                }
                else {
                    yield MastApi.makeCacheDir();
                    yield MastApi.writeFile(path, data, true);
                }
            }
        });
    }
    static readFile(path) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield new Promise((resolve, reject) => readFile(path, (err, data) => {
                if (err) {
                    reject(err);
                }
                else {
                    resolve(data.toString());
                }
            }));
        });
    }
    static downloadBundle(urlList, outputFile, progressTimer = 10000) {
        return __awaiter(this, void 0, void 0, function* () {
            const url = "https://mast.stsci.edu/api/v0.1/Download/bundle.zip";
            try {
                yield MastApi.makeDataDir();
            }
            catch (_a) { }
            const out = `${MastApi.DATA_DIR}/${outputFile}`;
            const writer = createWriteStream(out);
            const params = new URLSearchParams();
            console.log(`Getting ${urlList.length} files`);
            urlList.forEach((url) => params.append("uri", url));
            const response = yield yield axios.post(url, params.toString(), {
                responseType: "stream",
            });
            let size = 0;
            let done = false;
            const printSize = () => console.log(`Downloaded: ${size} bytes`);
            const printProgress = () => {
                if (!done) {
                    printSize();
                    setTimeout(printProgress, progressTimer);
                }
            };
            setTimeout(printProgress, progressTimer);
            yield new Promise((resolve, reject) => {
                let error;
                writer.on("error", (err) => {
                    error = err;
                    writer.close();
                    reject(err);
                });
                writer.on("close", () => {
                    if (!error) {
                        resolve();
                    }
                });
                writer.on("pipe", (chunk) => {
                    chunk.on("data", (chunk2) => {
                        size = size + chunk2.length;
                    });
                    chunk.on("close", () => {
                        done = true;
                        console.log(`Finished downloading ${urlList.length} files to ${out}`);
                    });
                });
                response.data.pipe(writer);
            });
        });
    }
    static filterByCollection(data, collection, filterExclusive = true) {
        const filtered = data.filter((entry) => {
            return (entry.obs_collection.includes(collection) &&
                (entry.dataRights !== "EXCLUSIVE_ACCESS" || !filterExclusive));
        });
        return filtered;
    }
    getObjectCache() {
        return __awaiter(this, void 0, void 0, function* () {
            if (!this.objectCache) {
                try {
                    const raw = yield MastApi.readFile(MastApi.OBJECT_CACHE);
                    if (raw) {
                        this.objectCache = JSON.parse(raw);
                    }
                    else {
                        this.objectCache = {};
                    }
                }
                catch (_a) {
                    this.objectCache = {};
                }
            }
            if (!this.objectCache) {
                throw new Error("Unable to load object cache");
            }
            return this.objectCache;
        });
    }
    setObjectCache(name, response) {
        return __awaiter(this, void 0, void 0, function* () {
            const cache = yield this.getObjectCache();
            cache[name] = response;
        });
    }
    getObservationCache() {
        return __awaiter(this, void 0, void 0, function* () {
            if (!this.observationCache) {
                try {
                    const raw = yield MastApi.readFile(MastApi.OBSERVATION_CACHE);
                    if (raw) {
                        this.observationCache = JSON.parse(raw);
                    }
                    else {
                        this.observationCache = {};
                    }
                }
                catch (_a) {
                    this.observationCache = {};
                }
            }
            if (!this.observationCache) {
                throw new Error("Unable to load observation cache");
            }
            return this.observationCache;
        });
    }
    setObservationCache(name, response) {
        return __awaiter(this, void 0, void 0, function* () {
            const cache = yield this.getObservationCache();
            cache[name] = response;
        });
    }
    getProductCache() {
        return __awaiter(this, void 0, void 0, function* () {
            if (!this.productCache) {
                try {
                    const raw = yield MastApi.readFile(MastApi.PRODUCT_CACHE);
                    if (raw) {
                        this.productCache = JSON.parse(raw);
                    }
                    else {
                        this.productCache = {};
                    }
                }
                catch (error) {
                    this.productCache = {};
                }
            }
            if (!this.productCache) {
                throw new Error("Unable to load product cache");
            }
            return this.productCache;
        });
    }
    setProductCache(obsId, result) {
        return __awaiter(this, void 0, void 0, function* () {
            const cache = yield this.getProductCache();
            cache[obsId] = result;
        });
    }
    // TODO Paging
    request(query) {
        return __awaiter(this, void 0, void 0, function* () {
            const searchParams = new URLSearchParams();
            if (!query.format) {
                query.format = RequestFormat.json;
            }
            const paramString = JSON.stringify(query);
            searchParams.append("request", paramString);
            try {
                const response = yield axios.post("https://mast.stsci.edu/api/v0/invoke", searchParams, {
                    headers: {
                        "Content-type": "application/x-www-form-urlencoded",
                        Accept: "text/plain",
                    },
                });
                return response.data;
            }
            catch (error) {
                console.error(`Query Error:/${JSON.stringify(query, null, 2)}`);
                throw error;
            }
        });
    }
    /**
     * https://mast.stsci.edu/api/v0/_services.html#MastNameLookup
     *
     * Resolves an object name into a position on the sky.
     * @param input (required) The object name to be resolved.
     * @returns
     */
    nameLookup(input, format = RequestFormat.json, useCache = true) {
        return __awaiter(this, void 0, void 0, function* () {
            const timerName = `Lookup ${input}`;
            console.time(timerName);
            const cache = (yield this.getObjectCache())[input];
            const service = Service.NameLookup;
            const params = { input, format };
            if (useCache && cache !== undefined) {
                console.timeEnd(timerName);
                return cache;
            }
            else {
                const result = yield this.request({
                    service,
                    params,
                });
                yield this.setObjectCache(input, result);
                console.timeEnd(timerName);
                return result;
            }
        });
    }
    /**
     * https://mast.stsci.edu/api/v0/_services.html#MastCaomCone
     *
     * Perform a CAOM cone search. See CAOM Field documentation (https://mast.stsci.edu/api/v0/_c_a_o_mfields.html) for the list of columns returned.
     *
     * @param ra Right ascension in decimal degrees.
     * @param dec Declination in decimal degrees.
     * @param radius Search radius in decimal degrees.
     * @param exclude_hla If true excludes HLA results
     */
    coneSearch(params) {
        return __awaiter(this, void 0, void 0, function* () {
            const timerName = `Cone search: ${JSON.stringify(params, null, 2)}`;
            console.time(timerName);
            const service = Service.ConeSearch;
            if (params.radius === undefined) {
                params.radius = 0.2;
            }
            if (params.exclude_hla === undefined) {
                params.exclude_hla = false;
            }
            const result = yield this.request({ service, params });
            console.timeEnd(timerName);
            const data = result.data;
            return data;
        });
    }
    filteredSearch(filters, count = false, format = RequestFormat.json) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = {
                service: Service.Filtered,
                format,
                params: {
                    columns: count ? "COUNT_BIG(*)" : "*",
                    filters,
                },
            };
            const result = yield this.request(request);
            return result;
        });
    }
    filteredPositionSearch(filters, coordinates, count = false, format = RequestFormat.json) {
        return __awaiter(this, void 0, void 0, function* () {
            const timerName = `Filtered`;
            const service = Service.FilteredPosition;
            const position = `${coordinates.ra}, ${coordinates.decl}, ${coordinates.radius || 0.2}`;
            const request = {
                service,
                format,
                params: {
                    columns: count ? "COUNT_BIG(*)" : "*",
                    filters,
                    position,
                },
            };
            console.time(timerName);
            const result = yield this.request(request);
            console.timeEnd(timerName);
            return result;
        });
    }
    getProduct(obsid) {
        return __awaiter(this, void 0, void 0, function* () {
            const timerString = `Get product: ${obsid}`;
            console.time(timerString);
            const cached = yield this.getProductCache();
            if (cached[obsid]) {
                console.timeEnd(timerString);
                return cached[obsid];
            }
            else {
                const service = Service.Products;
                const params = { obsid };
                const result = (yield this.request({
                    service,
                    params,
                }));
                yield this.setProductCache(obsid, result.data);
                console.timeEnd(timerString);
                return result.data;
            }
        });
    }
    getDownloadUrls(obsids) {
        return __awaiter(this, void 0, void 0, function* () {
            const urls = [];
            let size = 0;
            yield obsids.reduce((prev, obsid) => __awaiter(this, void 0, void 0, function* () {
                yield prev;
                const obs = yield this.getProduct(obsid);
                obs.forEach((observation) => {
                    urls.push(observation.dataURI);
                    size = size + observation.size;
                });
                console.log(`${obsid}: ${obs.length} entries ${size}`);
            }), Promise.resolve());
            return { size, urls };
        });
    }
}
/**
 * A cache of things resolved with nameLookup
 */
MastApi.CACHE_DIR = "./cache";
MastApi.DATA_DIR = "./data";
MastApi.OBJECT_CACHE = `${MastApi.CACHE_DIR}/objects.json`;
MastApi.OBSERVATION_CACHE = `${MastApi.CACHE_DIR}/observations.json`;
MastApi.PRODUCT_CACHE = `${MastApi.CACHE_DIR}/products.json`;
